libursa/src/errors/mod.rs

```diff
@@ -2,197 +2,217 @@
 extern crate serde_json;
 extern crate log;
 
-use std::error::Error;
-use std::{fmt, io};
+#[cfg(feature = "ffi")]
+use ffi::ErrorCode;
 
-#[derive(Debug, PartialEq, Copy, Clone, Serialize)]
-#[repr(usize)]
-pub enum ErrorCode
-{
-    Success = 0,
-
-    // Common errors
-
-    // Caller passed invalid value as param 1 (null, invalid json and etc..)
-    CommonInvalidParam1 = 100,
-
-    // Caller passed invalid value as param 2 (null, invalid json and etc..)
-    CommonInvalidParam2 = 101,
-
-    // Caller passed invalid value as param 3 (null, invalid json and etc..)
-    CommonInvalidParam3 = 102,
+use std::fmt;
+use std::cell::RefCell;
+#[cfg(feature = "ffi")]
+use std::ptr;
+use std::ffi::CString;
+#[cfg(feature = "ffi")]
+use std::os::raw::c_char;
 
-    // Caller passed invalid value as param 4 (null, invalid json and etc..)
-    CommonInvalidParam4 = 103,
+use failure::{Backtrace, Context, Fail};
 
-    // Caller passed invalid value as param 5 (null, invalid json and etc..)
-    CommonInvalidParam5 = 104,
+#[cfg(feature = "ffi")]
+use utils::ctypes;
 
-    // Caller passed invalid value as param 6 (null, invalid json and etc..)
-    CommonInvalidParam6 = 105,
-
-    // Caller passed invalid value as param 7 (null, invalid json and etc..)
-    CommonInvalidParam7 = 106,
-
-    // Caller passed invalid value as param 8 (null, invalid json and etc..)
-    CommonInvalidParam8 = 107,
-
-    // Caller passed invalid value as param 9 (null, invalid json and etc..)
-    CommonInvalidParam9 = 108,
-
-    // Caller passed invalid value as param 10 (null, invalid json and etc..)
-    CommonInvalidParam10 = 109,
+#[cfg(feature = "ffi")]
+pub mod prelude {
+    pub use super::{err_msg, UrsaCryptoError, UrsaCryptoErrorExt, UrsaCryptoErrorKind, UrsaCryptoResult, set_current_error, get_current_error_c_json};
+}
 
-    // Caller passed invalid value as param 11 (null, invalid json and etc..)
-    CommonInvalidParam11 = 110,
+#[cfg(not(feature = "ffi"))]
+pub mod prelude {
+    pub use super::{err_msg, UrsaCryptoError, UrsaCryptoErrorExt, UrsaCryptoErrorKind, UrsaCryptoResult};
+}
 
-    // Caller passed invalid value as param 11 (null, invalid json and etc..)
-    CommonInvalidParam12 = 111,
+#[derive(Copy, Clone, Eq, PartialEq, Debug, Fail)]
+pub enum UrsaCryptoErrorKind {
+    // Common errors
+    #[fail(display = "Invalid library state")]
+    InvalidState,
+    #[fail(display = "Invalid structure")]
+    InvalidStructure,
+    #[fail(display = "Invalid parameter {}", 0)]
+    InvalidParam(u32),
+    #[fail(display = "IO error")]
+    IOError,
+    // CL errors
+    #[fail(display = "Proof rejected")]
+    ProofRejected,
+    #[fail(display = "Revocation accumulator is full")]
+    RevocationAccumulatorIsFull,
```

Comment:
`RevocationAccumulatorIsFull` feels out of place with the rest of the new error types, perhaps this could be made more general? (e.g. in the same way `AnoncredsProofRejected` became `ProofRejected`)
`AccumulatorIsFull` would be better, but perhaps even something more generic could work.

Reply:
I didn't pick these the indy team did. This was just a matter of syncing recent changes in indy-crypto that weren't merged over.